include/swift/AST/DeclContext.h

```diff
@@ -255,6 +255,10 @@ class alignas(1 << DeclContextAlignInBits) DeclContext {
   /// EnumDecl, otherwise return null.
   EnumDecl *getAsEnumOrEnumExtensionContext() const;
 
+  /// If this DeclContext is a protocol and not an extension on a
+  /// protocol, return the ProtocolDecl, otherwise return null.
+  ProtocolDecl *getAsProtocolContext() const;
```

Comment:
The LLVM casting infrastructure already gives us a way to do this: `dyn_cast<ProtocolDecl>(declContext)`. So we don't need a new method.